snap/types.go

```diff
@@ -182,3 +182,38 @@ func (daemonScope *DaemonScope) fromString(str string) error {
 	*daemonScope = d
 	return nil
 }
+
+// ComponentType is a type to represent the possible types of snap components.
+type ComponentType string
+
+const (
+	// TestComponent is just for testing purposes.
+	// TODO add here new component when there is more progress on the
+	// components implementation.
+	TestComponent ComponentType = "test"
+)
+
+var validComponentTypes = [...]ComponentType{TestComponent}
+
+// Component represents a snap component.
+type Component struct {
+	Type        ComponentType `yaml:"type"`
+	Summary     string        `yaml:"summary"`
+	Description string        `yaml:"description"`
+}
+
+func (ct *ComponentType) UnmarshalYAML(unmarshall func(interface{}) error) error {
+	typeStr := ""
+	if err := unmarshall(&typeStr); err != nil {
+		return err
+	}
+
+	for _, t := range validComponentTypes {
```

Comment:
(super nitpick) I wonder if we could use `strutil.ListContains()` here? Something like:
```go
diff --git a/snap/types.go b/snap/types.go
index 1633cdcd35..9e689e494a 100644
--- a/snap/types.go
+++ b/snap/types.go
@@ -22,6 +22,8 @@ package snap
 import (
        "encoding/json"
        "fmt"
+
+       "github.com/snapcore/snapd/strutil"
 )
 
 // Type represents the kind of snap (app, core, gadget, os, kernel, snapd)
@@ -193,7 +195,7 @@ const (
        TestComponent ComponentType = "test"
 )
 
-var validComponentTypes = [...]ComponentType{TestComponent}
+var validComponentTypes = [...]string{string(TestComponent)}
 
 // Component represents a snap component.
 type Component struct {
@@ -208,11 +210,9 @@ func (ct *ComponentType) UnmarshalYAML(unmarshall func(interface{}) error) error
                return err
        }
 
-       for _, t := range validComponentTypes {
-               if string(t) == typeStr {
-                       *ct = t
-                       return nil
-               }
+       if strutil.ListContains(validComponentTypes[:], typeStr) {
+               *ct = ComponentType(typeStr)
+               return nil
        }
 
        return fmt.Errorf("unknown component type %q", typeStr)
```
but probably silly

Reply:
Actually with go1.21 and https://pkg.go.dev/slices@master#Contains this will also be nice, so let's just wait for 1.21 and keep as is :)
```go
if slices.Contains(validComponentTypes[:], ComponentType(typeStr)) {
	*ct = ComponentType(typeStr)
	return nil
}
```